File: go/ast/expr.go
package ast

import (
	"github.com/AsazuTaiga/crafting_interpriters/go/token"
)

type Expr interface {
	Accept(visitor Visitor) interface{}
}

type Visitor interface {
	VisitBinaryExpr(expr BinaryExpr) interface{}
	VisitGroupingExpr(expr GroupingExpr) interface{}
	VisitLiteralExpr(expr LiteralExpr) interface{}
	VisitUnaryExpr(expr UnaryExpr) interface{}
}

type BinaryExpr struct {
	Left Expr
	Operator token.Token
	Right Expr
}

func NewBinaryExpr(left Expr, operator token.Token, right Expr) *BinaryExpr {
	return &BinaryExpr{
		Left: left,
		Operator: operator,
		Right: right,
	}
}

func (expr *BinaryExpr) Accept(visitor Visitor) interface{} {
	return visitor.VisitBinaryExpr(*expr)
}

type GroupingExpr struct {
	Expression Expr
}

func NewGroupingExpr(expression Expr) *GroupingExpr {
	return &GroupingExpr{
		Expression: expression,
	}
}

func (expr *GroupingExpr) Accept(visitor Visitor) interface{} {
	return visitor.VisitGroupingExpr(*expr)
}

type LiteralExpr struct {
	Value interface{}
}

func NewLiteralExpr(value interface{}) *LiteralExpr {
	return &LiteralExpr{
		Value: value,
	}
}

func (expr *LiteralExpr) Accept(visitor Visitor) interface{} {
	return visitor.VisitLiteralExpr(*expr)
}

type UnaryExpr struct {
	Operator token.Token
	Right Expr
}

func NewUnaryExpr(operator token.Token, right Expr) *UnaryExpr {
	return &UnaryExpr{
		Operator: operator,
		Right: right,
	}
}

func (expr *UnaryExpr) Accept(visitor Visitor) interface{} {
	return visitor.VisitUnaryExpr(*expr)
}


type VariableExpr struct {
	Name token.Token
}

func NewVariableExpr(name token.Token) *VariableExpr {
	return &VariableExpr{
		Name: name,
	}
}

func (expr *VariableExpr) Accept(visitor Visitor) interface{} {
	return visitor.VisitVariableExpr(*expr)
}

File: rust/src/expr.rs
// https://craftinginterpreters.com/representing-code.html

use crate::token::{Token, TokenType};

/// literal | unary | binary| grouping ;
pub enum Expr {
    // expression operator expression
    Binary(Box<Expr>, TokenType, Box<Expr>),
    // "(" expression ")" ;
    Grouping(Box<Expr>),
    // ( "-" | "!" ) expression ;
    Unary(TokenType, Box<Expr>),
    // NUMBER | STRING | "true" | "false" | "nil" ;
    Literal(Literal),
}

#[derive(Debug)]
pub enum Literal {
    Number(f64),
    String(String),
    True,
    False,
    Nil,
}

trait Visitor {
    fn visit_binary_expr(&mut self, left: &Box<Expr>, binary_op: &TokenType, right: &Box<Expr>);
    fn visit_grouping_expr(&mut self, expr: &Expr);
    fn visit_literal_expr(&mut self, literal: &Literal);
    fn visit_unary_expr(&mut self, unary: &TokenType, expr: &Box<Expr>);
}

pub struct AstPrinter;

impl AstPrinter {
    pub fn print(&mut self, expr: &Expr) {
        expr.accept(self)
    }
}

impl Expr {
    fn accept(&self, visitor: &mut dyn Visitor) {
        match self {
            Expr::Binary(left, binary_op, right) => {
                visitor.visit_binary_expr(left, binary_op, right)
            }
            Expr::Grouping(expr) => visitor.visit_grouping_expr(expr),
            Expr::Literal(literal) => visitor.visit_literal_expr(literal),
            Expr::Unary(unary, expr) => visitor.visit_unary_expr(unary, expr),
        }
    }
}

impl Visitor for AstPrinter {
    fn visit_binary_expr(&mut self, left: &Box<Expr>, binary_op: &TokenType, right: &Box<Expr>) {
        print!("(");
        left.accept(self);
        match binary_op {
            TokenType::Minus => print!("-"),
            TokenType::Plus => print!("+"),
            TokenType::Slash => print!("/"),
            TokenType::Star => print!("*"),
            _ => {}
        }
        right.accept(self);
        print!(")");
    }
    fn visit_grouping_expr(&mut self, expr: &Expr) {
        print!("(");
        expr.accept(self);
        print!(")");
    }
    fn visit_literal_expr(&mut self, literal: &Literal) {
        match literal {
            Literal::Number(n) => print!("{}", n),
            Literal::String(s) => print!("{}", s),
            Literal::True => print!("true"),
            Literal::False => print!("false"),
            Literal::Nil => print!("nil"),
        }
    }
    fn visit_unary_expr(&mut self, unary: &TokenType, expr: &Box<Expr>) {
        print!("(");
        match unary {
            TokenType::Minus => print!("-"),
            TokenType::Bang => print!("!"),
            _ => {}
        }
        expr.accept(self);
        print!(")");
    }
}

fn main() {
    // -2 * (3 + 4)
    let tokens = Expr::Binary(
        Box::new(Expr::Unary(
            TokenType::Minus,
            Box::new(Expr::Literal(Literal::Number(2.0))),
        )),
        TokenType::Star,
        Box::new(Expr::Grouping(Box::new(Expr::Binary(
            Box::new(Expr::Literal(Literal::Number(3.0))),
            TokenType::Plus,
            Box::new(Expr::Literal(Literal::Number(4.0))),
        )))),
    );
    let mut printer = AstPrinter;
    printer.print(&tokens);
}
